src/lem/constrainer.rs

```diff
@@ -81,145 +117,320 @@ impl LEM {
-        }
         let alloc_ptr = Self::allocate_ptr(
-            &mut cs.namespace(|| format!("allocate input {}", &self.input[idx])),
-            &store.hydrate_ptr(&witness.input[idx])?,
+            cs,
+            &store.hydrate_ptr(&valuation.input[idx])?,
             &self.input[idx],
+            alloc_ptrs,
         )?;
         alloc_ptrs.insert(&self.input[idx], alloc_ptr.clone());
         Self::inputize_ptr(cs, &alloc_ptr, &self.input[idx])?;
         Ok(())
     }
 
-    /// Create R1CS constraints for LEM given an evaluation witness.
+    fn on_concrete_path(concrete_path: &Boolean) -> Result<bool> {
+        concrete_path
+            .get_value()
+            .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path"))
+    }
+
+    fn z_ptr_from_valuation<F: LurkField>(
+        concrete_path: &Boolean,
+        valuation: &Valuation<F>,
+        name: &String,
+        store: &mut Store<F>,
+    ) -> Result<ZPtr<F>> {
+        if Self::on_concrete_path(concrete_path)? {
+            let Some(ptr) = valuation.ptrs.get(name) else {
+                bail!("Couldn't retrieve {} from valuation", name);
+            };
+            store.hydrate_ptr(ptr)
+        } else {
+            Ok(ZPtr::dummy())
+        }
+    }
+
+    /// Accumulates slot data that will be used later to generate the constraints.
+    fn acc_slots_data_data<F: LurkField>(
+        concrete_path: Boolean,
+        slots_data: &mut SlotsData<F>,
+        img: AllocatedNum<F>,
+        img_name: String,
+        preimg_vec: Vec<&AllocatedPtr<F>>,
+    ) {
+        let arity = match preimg_vec.len() {
+            2 => HashArity::A2,
+            3 => HashArity::A3,
+            4 => HashArity::A4,
+            _ => unreachable!(),
+        };
+        let mut input_vec = Vec::new();
+        for elem in preimg_vec {
+            input_vec.push(elem.tag().clone());
+            input_vec.push(elem.hash().clone());
+        }
+        slots_data.push((arity, concrete_path, input_vec, img, img_name));
+    }
+
+    /// Use the implies logic to contrain tag and hash values for accumulated
+    /// slot data
+    fn create_slot_constraints<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        slots_data: &SlotsData<F>,
+        store: &mut Store<F>,
+        alloc_manager: &mut AllocationManager<F>,
+        num_hash_slots: &NumSlots,
+    ) -> Result<()> {
+        // Vectors fulls of dummies, so that it will not be required to fill with dummies later
+        let alloc_dummy_ptr = alloc_manager.get_or_alloc_ptr(cs, &ZPtr::dummy())?;
+        let mut hashes = vec![Some(alloc_dummy_ptr.hash().clone()); num_hash_slots.total()];
+
+        // In order to get a uniform circuit each type of hash has its own constant-size subvector
+        // Hash2 uses subvector from 0 to num_hash_slots.hash2
+        let mut hash2_index = 0;
+        // Hash3 uses subvector from num_hash_slots.hash2 to num_hash+slots.hash3
```

Comment:
'num_hash_slots', not 'num_hash+slots', unless I'm missing something subtle.

Reply:
I do think there's a subtlety here. I believe we're going from 0 to `num_hash_slots.hash2`, then from `num_hash_slots.hash2 + 1` to `num_hash_slots.hash2 + num_hash_slots.hash3` and then from `num_hash_slots.hash2 + num_hash_slots.hash3 + 1` to `num_hash_slots.total()`

Reply:
fixed

Reply:
I think it's incorrect though. See my comment above